Sprint 09

Fennec-meow opened the pull request.

Fennec-meow committed:
Submitting the 8th sprint for review.

Fennec-meow committed:
Parenthesis transfer.

Fennec-meow committed:
Correction

Fennec-meow committed:
sprint_09

Fennec-meow committed:
Alignment

So1ovyov commented on ImageFeed/ImagesList/ImagesListCell.swift:
```diff
@@ -0,0 +1,16 @@
+
+import UIKit
+
+final class ImagesListCell: UITableViewCell {
```
Хорошей практикой считается разделения разных блоков кода марками, например вот так:
```
final class ImagesListCell: UITableViewCell {

    // MARK: - Static properties

    static let reuseIdentifier = "ImagesListCell"

    // MARK: - @IBOutlet properties

    @IBOutlet weak var likeButton: UIButton!
    @IBOutlet weak var dateLabel: UILabel!
    @IBOutlet weak var cellImage: UIImageView!
}
```

So1ovyov commented on ImageFeed/ImagesList/ImagesListViewController.swift:
```diff
@@ -0,0 +1,100 @@
+
+import UIKit
+
+class ImagesListViewController: UIViewController {
```
Проверь во всех классах наличие модификатора `final`, он не нужен только если ты планируешь наследоваться от класса

So1ovyov commented on ImageFeed/ImagesList/ImagesListViewController.swift:
```diff
@@ -0,0 +1,100 @@
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        
+        tableView.rowHeight = 200
```
В данной настройке нет необходимости, так как высоту ячейки ты рассчитываешь ниже в методе `heightForRowAt`. Так что строку `tableView.rowHeight = 200` можно удалить.

So1ovyov commented on ImageFeed/ImagesList/ImagesListViewController.swift:
```diff
@@ -0,0 +1,100 @@
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = tableView.dequeueReusableCell(withIdentifier: ImagesListCell.reuseIdentifier, for: indexPath)
+        
+        guard let imageListCell = cell as? ImagesListCell else {
```
Обрати внимание, в кейсах с `guard` можно совершать сразу несколько действий, например создавать ячейку и кастить её:
```
        guard let imageListCell = tableView.dequeueReusableCell(
            withIdentifier: ImagesListCell.reuseIdentifier,
            for: indexPath
        ) as? ImagesListCell else {
            return UITableViewCell()
        }
```

So1ovyov commented on ImageFeed/ImagesList/ImagesListViewController.swift:
```diff
@@ -0,0 +1,100 @@
+        }
+        
+        cell.cellImage.image = image
+        cell.dateLabel.text = dateFormatter.string(from: Date())
```
Обрати внимание, каждый раз когда ты создаешь ячейку, в методе `configCell` объявляется новый экземпляр `Date()`. Например, если у тебя будет создано 1000 ячеек, то ты 1000 раз создаешь экземпляр `Date()`, согласись, это не оптимально. Предлагаю один раз создать свойство и использовать его во всех ячейках. Например так:
- Создаем свойство в классе
```
final class ImagesListViewController: UIViewController {
    private let currentDate = Date()
```
- Используем это свойство в методе `configCell`
```
cell.dateLabel.text = dateFormatter.string(from: currentDate)
```

So1ovyov commented on ImageFeed/Profile/ProfileViewController.swift:
```diff
@@ -0,0 +1,115 @@
+
+final class ProfileViewController: UIViewController {
+    
+    private var avatarImageView: UIImageView = {
```
Отличное решение!
Только внутри конструкции можно доп пустые строки не добавлять:
```
    private let avatarImageView: UIImageView = {
        let imageView = UIImageView()
        imageView.image = UIImage(named: "avatar")
        imageView.layer.cornerRadius = 35
        return imageView
    }()
```

So1ovyov replied:
Также у всех свойств будет достаточно константы `let`
```
private let avatarImageView: UIImageView = {
```

So1ovyov commented on ImageFeed/Profile/ProfileViewController.swift:
```diff
@@ -0,0 +1,115 @@
+    private var avatarImageView: UIImageView = {
+        let imageView = UIImageView()
+        
+        imageView.translatesAutoresizingMaskIntoConstraints = false
```
Строку `translatesAutoresizingMaskIntoConstraints = false` можно удалить и применить ниже в `forEach`

So1ovyov commented on ImageFeed/Profile/ProfileViewController.swift:
```diff
@@ -0,0 +1,115 @@
+    }
+    
+    private func setupContent() {
+        view.addSubview(avatarImageView)
```
Обрати внимание на инструмент `forEach` при верстке и настройке элементов, он поможет тебе сократить повторяющийся код:
```
    private func setupViews() {
        [avatarImageView, nameLabel, loginNameLabel, descriptionLabel, logoutButton].forEach {
            $0.translatesAutoresizingMaskIntoConstraints = false
            view.addSubview($0)
        }
    }
```

So1ovyov commented on ImageFeed/Profile/ProfileViewController.swift:
```diff
@@ -0,0 +1,115 @@
+    }
+    
+    private func setupConstraints() {
+        let avatarImageViewConstraints = [
```
Здесь можно с помощью инструмента `NSLayoutConstraints.activate` активировать сразу все констрейнты, а не каждый отдельно:
```
        private func setupСonstraints(){
            NSLayoutConstraint.activate([
                avatarImageView.widthAnchor.constraint(equalToConstant: 70),
                avatarImageView.heightAnchor.constraint(equalToConstant: 70),
                avatarImageView.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor, constant: 16),
                avatarImageView.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: 32),

                nameLabel.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor, constant: 16),
                nameLabel.topAnchor.constraint(equalTo: avatarImageView.bottomAnchor, constant: 8),
                nameLabel.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.trailingAnchor, constant: -16),

                /// и т.д.
            ])
        }
```

So1ovyov commented on ImageFeed/Profile/ProfileViewController.swift:
```diff
@@ -0,0 +1,115 @@
+    }
+    
+    @objc private func didTapLogoutButton() {
+        print("logout_button")
```
Предлагаю просто добавить `TODO` с описание что будет сделано в следующих итерациях:
```
    @objc
    private func didTapLogoutButton() {
        // TODO: - Добавить логику нажатия на кнопку Logout
    }
```

So1ovyov commented on ImageFeed/SingleImageView/SingleImageViewController.swift:
```diff
@@ -0,0 +1,74 @@
+        
+        return imageView
+    }
+}
```
Тут все отлично, только предлагаю не сильно зумить изображению, а расположить его по границам экрана, от левой до правой

Fennec-meow committed:
Changes:
1. The indentation between "SafeArea" and "tableView" is hidden, where the "Navigation Bar" is additionally located.
2. When you click on a cell, it is no longer highlighted in gray.
3. The "final" modifier has been added to the "ImagesListViewController" class.

So1ovyov commented on ImageFeed/Base.lproj/Main.storyboard:
```diff
@@ -21,15 +21,15 @@
                                             <subviews>
                                                 <imageView clipsSubviews="YES" userInteractionEnabled="NO" contentMode="scaleAspectFill" image="0" translatesAutoresizingMaskIntoConstraints="NO" id="r3b-hX-hrH">
-                                                    <rect key="frame" x="16" y="4" width="361" height="226"/>
+                                                    <rect key="frame" x="16" y="16" width="361" height="214"/>
```
Из-за того, что был добавлен отступ у ячейки с 4 на 16, между ячейками отступ теперь больше чем по дизайну. Это значение нужно вернуть к 4
А фикс заключался в отключении параметра `Shows Navigation Bar`,  что ты и сделала
![CleanShot 2025-01-18 at 12 44 35@2x](https://github.com/user-attachments/assets/bed09738-b6fe-401e-9aec-8e456c97bee0)

Fennec-meow committed:
Changes :
1. The margins between the cells have been changed, the old value "4" has been set.

So1ovyov commented:
@Fennec-meow все отлично, дизайн проекта полностью соответствует макету, критических замечаний нет.
Ты молодец, успехов в следующем спринте!

P.S. Если будет время и возможность, предлагаю применить в коде рекомендации оставленные выше в пул реквесте.
